### PR TITLE
resolving lock issues on deploy

### PR DESCRIPTION
**Pull request by thathert**

### CODE REVIEW

File: src/main/java/bio/terra/upgrade/Migrate.java
Comment (ddietterich) [inline src/main/java/bio/terra/upgrade/Migrate.java:56]: I'm curious about the larger system issue. It seems that K8s must provide some method for helping applications across the boundary of re-deploy. Are there either shutdown notifications (e.g. SIGINT or something) or startup hooks that could deterministically clear the lock?

This seems like a probabilistic approach that could lead to a whole new set of hard-to-diagnose failures.

Is there any case where we actually rely on the lock? Can we remove those cases? Seems to me that, in the current system, it is always the case that there is one DR Manager performing the migration and it can assume that it should own the lock. So the wait serves no purpose.

**Review by ddietterich — CHANGES_REQUESTED** [review]
I don't have confidence this is the right solution to the problem.

File: src/main/java/bio/terra/upgrade/Migrate.java
Comment (thathert) [reply src/main/java/bio/terra/upgrade/Migrate.java:56]: Good point. I had mentioned this in a previous PR and tried to capture the jist of the cause in a comment above.

Currently we rely on the lock when a new container is deploying because we will have two containers with potentially different migrations running at the same time. We would also rely on the lock when we start using deployments that have more than one API instance. The 5 minute timeout gives a container enough time to get the migration done and release the lock. If we hit the 5 minute mark then we should be able to assume that the container got killed so we would want the new container to clean up.  

I think this is a "band-aid" solution, and that we should look to a longer term solution that avoids this hackiness around the timeout.

There are startup hooks provided by k8s in the form of [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/). These must run to completion successfully before a pod is deployed. There are two approaches we can take with this:

1. Have the init container run the latest migration for the code that is to be deployed by an api container. This let's us move the migration out of the code and have it be tied to this lifecycle event. We would still need to handle the case of performing a rollback if a migration ran correctly in the init container and the api pod deployment failed, but it feels like the best long term solution.
2. Have the init container clean up locks before the api pod deploys. Gives us more determinism, but still leaves the case where a container running a migration gets killed and leaves us in a dead-locked state while the other containers wait for a release. I think we'd still need to leave the lock cleanup code in there..

So I think we should go with route 1 above. If we agree, I'll make a ticket for this and can pick this up next sprint. I'd still like to merge this PR in the short-term to unblock all of these PRs and make integration testing more fault tolerant.

Comment (ddietterich) [reply src/main/java/bio/terra/upgrade/Migrate.java:56]: Can you explain more about when there are two migrations running at the same time?
I thought the sequence was:
1. kill current DR manager - maybe leaving stuck locks
2. start new DR manager
Which can be solved by new DR manager saying "tally ho!" to the locks. Having it wait 5 minutes doesn't gain us anything.

For the larger problem, I'm not sure what the requirements are. Do we need to support continuous operation with rolling upgrades? If so, we will need code that can run against version A of the schema and then, on command, run against version B of the schema. If there is more than one DR manager, it requires synchronizing them on the migration action.

If that isn't a requirement, then we can simplify things. Take some downtime so we know that it is safe to seize the liquibase locks and do the migration.

One motivation for coding migration into the DR manager was to enable rolling upgrade support by separating the migration from the deployment.

Another concern I have related to migration is flights that need to be recovered by stairway. Will we have cases where we need to allow recovery to complete before migrating the stairway database format? Or are we confident we can always code around that?

A not-so-related concern is that we rely on the serialization/deserialization of objects stored in JSON by stairway. If the objects change across a version, then recovery may fail. That is on us to understand the full picture of version dependencies. I think it underscores the need to get some product input in this area and do some system design work.

Comment (thathert) [reply src/main/java/bio/terra/upgrade/Migrate.java:56]: Yes, so the issue I'm concerned about is when we get to production and we have one API deployment which will have more than one instances of the DR manager container running at the same time. Let's consider two containers, A and B.

1. A starts up and acquires a lock. It begins running through the changesets and migrating tables.
2. B starts up, sees that there is a lock, "tally ho!"s the lock and starts migrating.

A and B will be running through migrations at the same time. I don't think we have any guarantees from Liquibase at this point that the database will be left in a consistent state.

Another bad situation, let's say that we have a new version of DR manager (container Y) to deploy that depends on a migration that is not compatible with the currently deployed version (container X).

0. X has been running for a while, no migrations recently.
1. Y starts up and acquires a lock. It begins running through migrations. Y will not start getting traffic and X will not be shut down until Y successfully migrates and starts up.
2. The new migrations are incompatible with X. X starts failing so k8s reboots it.
3. New version of X starts up. X sees the lock from Y, releases it, and starts migrating.

We can again end up in a case where two migrations are running at the same time, and this time it is two different versions of the migration. The timeout should allow the container that acquired the lock to complete the migration and release the lock..

I agree with this: Take some downtime so we know that it is safe to seize the liquibase locks and do the migration.

I don't think we need to support rolling updates initially

Comment (ddietterich) [reply src/main/java/bio/terra/upgrade/Migrate.java:56]: Maybe I'm being dense... Right now, do we have any case where we have two DR manager containers starting at the same time? I'm just trying to understand whether your band-aid makes sense. It doesn't make sense to me if all we have is stuck locks from dead DR Managers.

As your cases illustrate, when we do have two DR managers running, we need some more complex coordination. I don't think we have that case right now. And I could be totally wrong and we do...

Comment (thathert) [reply src/main/java/bio/terra/upgrade/Migrate.java:56]: Yes, the second case from my last comment can and does happen right now. Kubernetes will not get rid of a container until it knows that the replacement container is able to start up and function. If the replacement container puts the database into a state that is not compatible with the original container, but there are issues with the replacement, then we can end up with two broken containers that are trying to start back up at the same time, failing, and then trying again in a little while.

This is unlikely to happen in a dev environment where we expect each container to be stable-ish, but it happens frequently in integration. One improvement is to always have the container dropAll before migrating, this at least solves the case where another container has put the database into a state where Liquibase complains about checksums left from a previous incompatible migration.

There is still the risk of a container being killed in the middle of a migration and leaving a lock that won't get released. Even with the "dropAll every time" approach, Liquibase will see the lock and patiently wait until it is released. This is why I'd like to add the lock release bandaid. 

You are right about the timeout probably being unnecessary while we are in the 1-dr-manager-at-a-time state if we treat this as a temporary band-aid while we come up with a better solution around coordinating multiple managers. I'll remove it if we agree on this path forward.

Comment (ddietterich) [reply src/main/java/bio/terra/upgrade/Migrate.java:56]: OK, got it. Thanks for your patience!

Maybe when I'm back we can work with Ruchi and Claire to get a good statement of our availability goals. With that, we can do some work to understand the range of cases we need to handle and figure out a design.

**Review by ddietterich — APPROVED** [review]

File: src/main/java/bio/terra/upgrade/Migrate.java
Comment (thathert) [reply src/main/java/bio/terra/upgrade/Migrate.java:56]: NP! Thanks for raising these great points. Definitely, I'll put something down for us to discuss when you're back in town.

**Review by RoriCremer — APPROVED** [review]

**Review by ahaessly — APPROVED** [review]
i have some questions i'll ask in faces

**Pull request merged**
